Source/Global/perform_env.cpp

```diff
@@ -294,8 +307,175 @@ HRESULT CALLBACK HC_PERFORM_ENV::HttpPerformAsyncShimProvider(XAsyncOp op, const
+    _Inout_ XAsyncBlock* asyncBlock
+)
+{
+    auto context = http_allocate_unique<WebSocketConnectContext>(this, std::move(uri), std::move(subprotocol), clientWebSocketHandle, asyncBlock);
+    RETURN_IF_FAILED(XAsyncBegin(asyncBlock, context.get(), (void*)HCWebSocketConnectAsync, nullptr, WebSocketConnectAsyncShimProvider));
+    context.release();
+
+    return S_OK;
+}
+
+HRESULT CALLBACK HC_PERFORM_ENV::WebSocketConnectAsyncShimProvider(XAsyncOp op, const XAsyncProviderData* data)
+{
+    WebSocketConnectContext* context{ static_cast<WebSocketConnectContext*>(data->context) };
+    HC_PERFORM_ENV* env{ context->env };
+
+    switch (op)
+    {
+    case XAsyncOp::Begin:
+    {
+        XTaskQueuePortHandle workPort{};
+        RETURN_IF_FAILED(XTaskQueueGetPort(data->async->queue, XTaskQueuePort::Work, &workPort));
+        RETURN_IF_FAILED(XTaskQueueCreateComposite(workPort, workPort, &context->internalAsyncBlock.queue));
+
+        std::unique_lock<std::mutex> lock{ env->m_mutex };
+        env->m_connectingWebSockets.insert(context->clientAsyncBlock);
+        lock.unlock();
+
+        return context->websocket->ConnectAsync(std::move(context->uri), std::move(context->subprotocol), &context->internalAsyncBlock);
+    }
+    case XAsyncOp::GetResult:
+    {
+        WebSocketCompletionResult* result{ reinterpret_cast<WebSocketCompletionResult*>(data->buffer) };
+        *result = context->connectResult;
+        return S_OK;
+    }
+    case XAsyncOp::Cleanup:
+    {
+        HC_UNIQUE_PTR<WebSocketConnectContext> reclaim{ context };
+        return S_OK;
+    }
+    default:
+    {
+        return S_OK;
+    }
+    }
+}
+
+void CALLBACK HC_PERFORM_ENV::WebSocketConnectComplete(XAsyncBlock* async)
+{
+    WebSocketConnectContext* context{ static_cast<WebSocketConnectContext*>(async->context) };
+    HC_PERFORM_ENV* env{ context->env };
+
+    std::unique_lock<std::mutex> lock{ env->m_mutex };
+    env->m_connectingWebSockets.erase(context->clientAsyncBlock);
+
+    // If cleanup is pending and the connect succeeded, immediately disconnect
+    bool disconnect{ false };
+
+    HRESULT hr = HCGetWebSocketConnectResult(&context->internalAsyncBlock, &context->connectResult);
+    if (SUCCEEDED(hr) && SUCCEEDED(context->connectResult.errorCode))
+    {
+        env->m_connectedWebSockets.insert(new (http_stl_allocator<ActiveWebSocketContext>{}.allocate(1)) ActiveWebSocketContext{ env, context->websocket });
+        if (env->m_cleanupAsyncBlock)
+        {
+            disconnect = true;
+        }
+    } 
+    
+    bool scheduleProviderCleanup = env->CanScheduleProviderCleanup();
+    lock.unlock();
+
+    assert(!scheduleProviderCleanup || !disconnect);
+    if (disconnect)
+    {
+        hr = context->websocket->Disconnect();
+        if (FAILED(hr))
+        {
+            HC_TRACE_ERROR_HR(HTTPCLIENT, hr, "WebSocket::Disconnect failed during HCCleanup");
+        }
+    }
+    else if (scheduleProviderCleanup)
+    {
```

Comment:
i see the assert but this might be a little safer if this was just else {} instead of else if {}.

Reply:
I actually think "else if" is safer. If somehow the assertion is false and both 'disconnect' and 'scheduleProviderCleanup' are true, ProviderCleanup would scheduled twice - once here and once when disconnect completes.

Reply:
I don't follow that logic at all:
the assert says there's 3 valid paths through this code:
1. only `disconnect == true`
2. only `scheduleProviderCleanup == true`
3. neither is true

(ie at least one of `disconnect`, `scheduleProviderCleanup` must be false)

Changing the `else if` into and `if` collapses case 2 and 3 into going into the branch which is obviously a massive and incorrect behavior change.

But your comment implies that at least one of them must be true, which we are not asserting at all right now, so which is it?

Reply:
Maybe my comment was unclear.

As the assertion implies at least one of these must be false.  All WebSockets need to be disconnected before cleanup happens.  If 'disconnect' is true, scheduleProviderCleanup should be false.  

What I was trying to convey in my last comment was that if somehow both of these ended up true (due to a bug somewhere) keeping it as an 'else if' protects against both paths executing, leading to ProviderCleanup being schedule twice.

Reply:
both `else` and `else if` do that
they will only differ is both flags are false

Reply:
Doh....I read Jason's original comment wrong.  I thought he was suggesting changing the 'else if' to just 'if'.

'else' definitely doesn't make sense.  It is a perfectly valid scenario for them to both be false.

Reply:
cool cool